src/pybind/mgr/volumes/fs/operations/trash.py

```diff
@@ -53,37 +55,75 @@ def get_trash_entry(self, exclude_list):
-                    d = self.fs.readdir(dir_handle)
-                    while d and not should_cancel():
-                        if d.d_name not in (b".", b".."):
-                            d_full = os.path.join(root_path, d.d_name)
-                            if d.is_dir():
-                                rmtree(d_full)
-                            else:
-                                self.fs.unlink(d_full)
-                        d = self.fs.readdir(dir_handle)
-            except cephfs.ObjectNotFound:
-                return
-            except cephfs.Error as e:
-                raise VolumeException(-e.args[0], e.args[1])
-            # remove the directory only if we were not asked to cancel
-            # (else we would fail to remove this anyway)
-            if not should_cancel():
-                self.fs.rmdir(root_path)
-
-        # catch any unlink errors
+        log.debug(f'purge(): trashpath = {trashpath}')
+
+        # XXX: add trailing '/' since its absence will produce relative path
+        # with leading '/' which will misguide os.path.join() to produce
+        # unexpected results.
+        #
+        # os.path.join('abcd', '123') = '/abcd/123'   --> expected
+        # os.path.join('abcd', '/123') = '/123'       --> unexpected
+        if trashpath.decode()[len(trashpath) - 1] != '/':
+            trashpath += b'/'
+
         try:
-            rmtree(trashpath)
+            # each stack member is tuple containing path to a directory and
+            # handle for that directory.
+            #
+            # The path here is a relative path (relative to trash path) instead
+            # of absolute path. This is to minimize the memory consumption and
+            # thereby also minimize the chance of running out of memory.
+            #
+            # The directory handle is also stored on stack to avoid unnecessary
+            # repetitive calls to fs.opendir().
+            stack = deque([(b'', self.fs.opendir(trashpath)), ])
+            while stack and not should_cancel():
+                log.info(f'here123 1')
+                curr_dir_path = os.path.join(trashpath, stack[len(stack) - 1][0])
+                curr_dir_handle = stack[len(stack) - 1][1]
+
+                # de = directory entry
+                de = self.fs.readdir(curr_dir_handle)
+                while de and not should_cancel():
+                    log.info(f'here123 2')
+                    if de.d_name in (b'.', b'..'):
+                        de = self.fs.readdir(curr_dir_handle)
+                        continue
+
+                    de_path = os.path.join(curr_dir_path, de.d_name)
+                    # optimization: if directory is empty, avoid pushing it on
+                    # stack and then deleting it and popping it from stack,
+                    # since it can be deleted right away.
+                    if de.is_dir():
+                        if is_dir_empty(self.fs, de_path):
```

Comment:
this call to `is_dir_empty` should be avoided: keep everything in this method.

Reply:
That would make this method too bloated IMO. Therefore, I've added a `is_dir_empty()` within this method which doesn't use any helper method and this method is being called instead.